### PR TITLE
cmd/compile: fix errors in zbs rules

### PR DESCRIPTION
**Pull request by lxq015**
Fix several errors in the Zbs instruction rules.

Update #73

### CODE REVIEW

**Review by wangpc-pp — COMMENTED** [review]
Please take a look if this fixes the error we met. @BoyaoWang430

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules]: `(x >> c) & mask` cannot be rewritten directly to `BEXTI` in value-producing contexts. The original expression returns `0` or `mask`, while `BEXTI` returns `0` or `1`. This is only equivalent when `mask == 1`. For `mask != 1`, the extracted bit must be shifted back to preserve the original value.

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules [hunk 1]
```diff
@@ -101,16 +107,8 @@
 (SEQZ (AND x (MOVDconst [c]))) && oneBit64(c) && log64(c) >= 10 && buildcfg.GORISCV64 >= 22 => (SEQZ (BEXTI <typ.UInt64> [log64(c)] x))
 
 // Pattern: Direct ANDI with shifted value (rare but possible)
-(SNEZ (ANDI (SRL x y) [1])) && buildcfg.GORISCV64 >= 22 => (BEXT x y)
-(SNEZ (ANDI (SRLW x y) [1])) && buildcfg.GORISCV64 >= 22 => (BEXT x y)
-(SNEZ (ANDI (SRLI x [c]) [1])) && c < 64 && buildcfg.GORISCV64 >= 22 => (BEXTI [c] x)
-(SNEZ (ANDI (SRLIW x [c]) [1])) && c < 32 && buildcfg.GORISCV64 >= 22 => (BEXTI [c] x)
 (SNEZ (ANDI (SLLI x [c]) [1])) && c > 0 && c < 64 && buildcfg.GORISCV64 >= 22 => (BEXTI [64-c] x)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules]: They are not valid for c > 0. In fixed-width integer arithmetic, (x << c) & 1 is always zero when c > 0, because left shift fills low bits with zero. It does not test bit 64-c.

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules:66]: SRLW uses y&31 as the shift amount, while BEXT on RV64 uses y&63 as the bit index. Check all SRLW

Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules]: This is not correct for `c == 31`. On RV64, `SLLIW [31] (MOVDconst [1])` sign-extends the 32-bit result and produces `0xffffffff80000000`, not the 64-bit single-bit mask `0x0000000080000000`. Please check all SLLIW

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules [hunk 2]
```diff
@@ -53,6 +53,11 @@
 // The compiler inserts AND with mask to handle out-of-range shifts (e.g., y >= width)
 // We optimize the bit extraction part, keeping the mask for correctness
 // NOTE: This rule must come before the simpler (ANDI (SRL x y) [1]) pattern to match first
+// When mask is even (mask&1==0), the result is always 0
+(ANDI (AND (SRL x y) (MOVDconst [mask])) [1]) && (mask&1)==0 => (MOVDconst [0])
+(ANDI (AND (SRL x y) (MOVDconst [mask])) [1]) && (mask&1)==1 => (BEXT x y)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules]: Add RVA22 guard

**Review by BoyaoWang430 — COMMENTED** [review]
Please check carefully. Maybe also add some test.

**Comment by lxq015** [comment]
> Please check carefully. Maybe also add some test.

@BoyaoWang430 Thanks for your careful review! I'm sorry about these errors. I've gone through every Zbs rule and fixed the issues you identified. Some redundant or incorrect rules were removed, and additional tests were added for the boundary cases. Please take another look when you have a chance. Thanks again!

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules [hunk 3]
```diff
@@ -62,90 +66,76 @@
...
-(SNEZ (AND x (SLLW (MOVDconst [1]) y))) && buildcfg.GORISCV64 >= 22 => (BEXT x y)
 (SNEZ (AND x (SLLI [c] (MOVDconst [1])))) && c < 64 && buildcfg.GORISCV64 >= 22 => (BEXTI [c] x)
-(SNEZ (AND x (SLLIW [c] (MOVDconst [1])))) && c < 32 && buildcfg.GORISCV64 >= 22 => (BEXTI [c] x)
-(SNEZ (AND x (MOVDconst [c]))) && oneBit64(c) && log64(c) >= 10 && buildcfg.GORISCV64 >= 22 => (BEXTI [log64(c)] x)
+(SNEZ (AND x (SLLIW [c] (MOVDconst [1])))) && c < 31 && buildcfg.GORISCV64 >= 22 => (BEXTI [c] x)
+(SNEZ (AND x (MOVDconst [c]))) && isUint64PowerOfTwo(c) && log64(c) >= 10 && buildcfg.GORISCV64 >= 22 => (BEXTI [log64(c)] x)
 
 // Pattern: 32-bit wrapped patterns for BEXT
+// Note: No boundary check on y - hardware handles it correctly
 (SNEZ (MOVWUreg (AND x (SLL (MOVDconst [1]) y)))) && buildcfg.GORISCV64 >= 22 => (BEXT x y)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules]: This is not correct because MOVWU will truncate to low 32 bits while BEXT may extract higher bit (y >= 32). The same for Line93

File: src/cmd/compile/internal/ssa/_gen/RISCV64.rules [hunk 4]
```diff
@@ -538,31 +538,30 @@
...
 
 // Pattern: AND with constant left shift (x & (1 << c))
 // Matches: if x&(1<<c) != 0 where c is a constant
 (B(EQ|NE)Z (AND x (SLLI [c] (MOVDconst [1]))) yes no) && c < 64 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
-(B(EQ|NE)Z (AND x (SLLIW [c] (MOVDconst [1]))) yes no) && c < 32 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
+(B(EQ|NE)Z (AND x (SLLIW [c] (MOVDconst [1]))) yes no) && c < 31 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
 
 // Pattern: 32-bit bit test with variable shift wrapped in MOVWUreg/MOVWreg
 // Matches: if uint32(x)&(1<<y) != 0 or int32(x)&(1<<y) != 0
 (B(EQ|NE)Z (MOVWUreg (AND x (SLL (MOVDconst [1]) y))) yes no) && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXT <typ.UInt64> x y) yes no)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64.rules]: This is not correct because MOVWU will truncate to low 32 bits while BEXT may extract higher bit (y >= 32). The same for line556

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules
Comment (BoyaoWang430) [reply src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules:66]: Line61 and Line66 still not correct

File: src/cmd/compile/internal/ssa/_gen/RISCV64.rules [hunk 5]
```diff
@@ -576,17 +575,16 @@
 (B(EQ|NE)Z (ANDI [1] (SRLW x y)) yes no) && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXT <typ.UInt64> x y) yes no)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64.rules:575]: SRLW issue

File: src/cmd/compile/internal/ssa/_gen/RISCV64.rules [hunk 6]
```diff
@@ -576,17 +575,16 @@
...
 
 // Pattern: 32-bit bit test patterns with MOVWUreg wrapper
-(B(EQ|NE)Z (MOVWUreg (SRLIW [c] x)) yes no) && c < 32 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
-(B(EQ|NE)Z (MOVWUreg (ANDI [1] (SRLI [c] x))) yes no) && c < 64 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
+// Note: For SRLI (64-bit shift) wrapped in MOVWUreg, only bits 0-31 are meaningful
+(B(EQ|NE)Z (MOVWUreg (ANDI [1] (SRLI [c] x))) yes no) && c < 32 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
 (B(EQ|NE)Z (MOVWUreg (ANDI [1] (SRLIW [c] x))) yes no) && c < 32 && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXTI <typ.UInt64> [c] x) yes no)
 (B(EQ|NE)Z (MOVWUreg (ANDI [1] (SRL x y))) yes no) && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXT <typ.UInt64> x y) yes no)
 (B(EQ|NE)Z (MOVWUreg (ANDI [1] (SRLW x y))) yes no) && buildcfg.GORISCV64 >= 22 => (B(EQ|NE)Z (BEXT <typ.UInt64> x y) yes no)
```
Comment (BoyaoWang430) [inline src/cmd/compile/internal/ssa/_gen/RISCV64.rules:582]: SRLW issue

File: src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules
Comment (lxq015) [reply src/cmd/compile/internal/ssa/_gen/RISCV64latelower.rules:66]: Thanks for your patience and time.I kept these SRLW→BEXT rules because the lowering guarantees y < 32 before bare SRLW can appear:
```
// Unbounded shift: produces masked SRLW, won't match bare SRLW patterns
(Rsh32Ux(8|16|32|64) x y) && !shiftIsBounded(v) =>
    (AND (SRLW x y) (Neg32 (SLTIU [32] y)))

// Bounded shift: bare SRLW only when y < 32
(Rsh32Ux(64|32|16|8) x y) && shiftIsBounded(v) =>
    (SRLW x y)
```
This means:

If y >= 32 (unbounded) →  lowering produces (AND (SRLW x y) (Neg32 (SLTIU [32] y)))→ only matches the masked rule, which correctly preserves the bound-check mask
I also ran a test (func bit32_srlw in bits.go) to verify the actual codegen for the unbounded case.
The interesting result is that for `uint32 x, uint32 y`, the lowering actually
uses SRL (64-bit) rather than SRLW:
```
  v4  = SRL <uint64> v7 v8       // x >> y  (SRL, not SRLW!)
  v10 = AND <uint64> v4 v5       // bound-check mask
  v12 = ANDI <uint64> [1] v10    // & 1
```
This matches the `(ANDI (AND (SRL x y) mask) [1])` rule in latelower,
producing `BEXT x, y` followed by `AND mask`:
```
  BEXT  X11, X10, X6    // extract bit y from x
  AND   X5, X6, X5      // preserve bound-check mask
```
When y >= 64, mask = 0, the AND zeros the output. When 32 <= y < 64,
the mask keeps BEXT's result, and since x is a zero-extended uint32,
bits 32-63 are always 0 — so BEXT is correct.

So all SRLW/SRL→BEXT rules are safe as-is.

If there's something I missed, please don't hesitate to let me know.

File: src/cmd/compile/internal/ssa/_gen/RISCV64.rules (hunk 4 above)
Comment (lxq015) [reply src/cmd/compile/internal/ssa/_gen/RISCV64.rules]: Sorry, I wrongly believed that these MOVWUreg/MOVWreg BEXT rules were triggered only when x and y were int32/uint32 type (where the upper bits are naturally masked). 

I have fixed this by adding an ANDI [c] && c >= 0 && c <= 31 guard to the rule patterns, ensuring BEXT is only generated when the shift amount is explicitly masked to [0, 31]. all.bash has passed.

**Review by BoyaoWang430 — APPROVED** [review]
+1

**Review by wangpc-pp — APPROVED** [review]
+2

**Review by wenchangping — APPROVED** [review]
+1

**Review by newborn22 — APPROVED** [review]
+2

**Review by ctk-1998 — APPROVED** [review]
+1
